### PR TITLE
FIX resolve previewing pages which redirect (#143)

### PR DESCRIPTION
**Pull request by wilr**
Issue https://github.com/silverstripe/silverstripe-sharedraftcontent/issues/143

This commit ensures that the Environment::getEnv('REQUEST_URI') value is
correctly updated when performing the internal redirection.

### CODE REVIEW

**Comment by michalkleiner** [comment]
@maxime-rainville or @emteknetnz, could you guys have a look here and/or tag someone who can? This is an issue for a big CWP site so would be good to get someone have a look over it if possible :) Thanks heaps!

File: src/Controllers/ShareDraftController.php [hunk 1]
```diff
@@ -119,18 +120,26 @@ public function preview(HTTPRequest $request)
      *
      * @return HTTPResponse
      */
-    protected function getRenderedPageByURLSegment($url)
```
Comment (emteknetnz) [inline src/Controllers/ShareDraftController.php:122]: Removing this protected function is technically a semver violation

Keep the existing getRenderedPageByURLSegment function, add a deprecation comment and a note that this does work with redirector pages

Then create a new private function for getRenderedPageByURL

File: src/Controllers/ShareDraftController.php [hunk 2]
```diff
@@ -119,18 +120,26 @@ public function preview(HTTPRequest $request)
...
+    protected function getRenderedPageByURL($url)
     {
-        $pageRequest = HTTPRequestBuilder::createFromEnvironment();
-        $pageRequest->setHttpMethod('GET');
-        $pageRequest->setUrl($url);
+        // Clean and update live global variables
+        $variables = HTTPRequestBuilder::cleanEnvironment(Environment::getVariables());
```
Comment (emteknetnz) [inline src/Controllers/ShareDraftController.php:157]: Put in a comment that you're basically doing what happens in `HTTPRequestBuilder::createFromEnvironment()`, I was pretty confused by the `@file_get_contents('php://input')` at first

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: src/Controllers/ShareDraftController.php (hunk 1 above)
Comment (wilr) [reply src/Controllers/ShareDraftController.php:122]: @emteknetnz Done.

File: src/Controllers/ShareDraftController.php (hunk 2 above)
Comment (wilr) [reply src/Controllers/ShareDraftController.php:157]: @emteknetnz Done.

**Comment by emteknetnz** [comment]
Failing build is codecov uploader related, nothing to do with this PR, I've raised an issue on their repo for this https://github.com/codecov/uploader/issues/166

File: src/Controllers/ShareDraftController.php [hunk 3]
```diff
@@ -117,20 +117,59 @@ public function preview(HTTPRequest $request)
...
+
+        return $response;
+    }
+
+    /**
+     * @param string $url
+     *
+     * @return HTTPResponse
+     */
+    private function getRenderedPageByURL($url)
```
Comment (emteknetnz) [inline src/Controllers/ShareDraftController.php]:
```suggestion
    private function getRenderedPageByURL(string $url): HTTPResponse
```

Up to top you'll also need to add
```
use SilverStripe\Control\HTTPResponse;
```

File: src/Controllers/ShareDraftController.php [hunk 4]
```diff
@@ -117,20 +117,59 @@ public function preview(HTTPRequest $request)
     /**
      * @param string $url
      *
+     * @deprecated Use {@link getRenderedPageByURL()}
```
Comment (emteknetnz) [inline src/Controllers/ShareDraftController.php]:
```suggestion
     * @deprecated 3.0 Use {@link getRenderedPageByURL()}
```

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Have tested locally with `axllent/silverstripe-trailing-slash 2.2.5` installed

Have also replicated the original issue with `axllent/silverstripe-trailing-slash 2.2.5` installed

Couple more minor changes requested

Also please retarget (and possibly rebase) this to `2.4` since we can release this as a patch rather than a new minor

**Comment by wilr** [comment]
@emteknetnz updated.

**Comment by emteknetnz** [comment]
@wilr Look good, I've retargetted the PR to 2.4 - though it's now got some extra commits in there that should be there - could you please rebase on 2.4

**Comment by wilr** [comment]
@emteknetnz done.

**Review by emteknetnz — APPROVED** [review]

**Comment by emteknetnz** [comment]
@wilr Merged

Thanks for the pull-request, I've tagged a patch release for you

https://github.com/silverstripe/silverstripe-sharedraftcontent/releases/tag/2.4.1

**Comment by michalkleiner** [comment]
Thanks Will, glad I asked you to keep digging into this instead of settling for a workaround ;-)

**Pull request merged**